Test Coverage Check for PR's

sandeep-v1404 opened the pull request:
**What kind of change does this PR introduce?**

github-actions

**Issue Number:**

Fixes #419  

**Did you add tests for your changes?**

No.

**Snapshots/Videos:**

![image](https://user-images.githubusercontent.com/57038543/148215457-12bf41d4-53f1-41bb-9d7a-27f46c4ac280.png)
![image](https://user-images.githubusercontent.com/57038543/148216235-79f92629-e11d-4695-ad5d-9c493b5ee0d2.png)
![image](https://user-images.githubusercontent.com/57038543/148217156-2740b03c-043e-4ae6-be43-dd2f641d391f.png)


**If relevant, did you update the documentation?**

**Summary**

- To Add Test Coverage Check for Pull Requests using GitHub Action
- To generate insights on tests when PR's are created and Commits have been pushed. 
- To fix issue #419  

**Does this PR introduce a breaking change?**

No

**Other information**

- Setup [Action secret](https://github.com/PalisadoesFoundation/talawa-api/settings/secrets/actions) with Name as `CODECOV_TOKEN` and value generated from [codecov](https://about.codecov.io/)
- We can get the insights After any PR or Push Commits on the repository. 
- Replace the Markdown badge in README.md which can be copied from settings>badge. 

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

palisadoes commented on README.md:
```diff
@@ -3,6 +3,7 @@
 [![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
 [![GitHub stars](https://img.shields.io/github/stars/PalisadoesFoundation/talawa-api.svg?style=social&label=Star&maxAge=2592000)](https://github.com/PalisadoesFoundation/talawa-api)
 [![GitHub forks](https://img.shields.io/github/forks/PalisadoesFoundation/talawa-api.svg?style=social&label=Fork&maxAge=2592000)](https://github.com/PalisadoesFoundation/talawa-api)
+[![codecov](https://codecov.io/gh/sandeep-v1404/talawa-api/branch/test-coverage-check/graph/badge.svg?token=LZ0LY3XQJ8)](https://codecov.io/gh/sandeep-v1404/talawa-api)
```
This needs to be changed to our badge

[![codecov](https://codecov.io/gh/PalisadoesFoundation/talawa-admin/branch/develop/graph/badge.svg?token=II0R0RREES)](https://codecov.io/gh/PalisadoesFoundation/talawa-admin)

sandeep-v1404 replied:
Sir, this needs the authorization to get the badge code.
![image](https://user-images.githubusercontent.com/57038543/148376560-eebd542d-d006-4e52-97c2-dd8ee3980c3a.png)

palisadoes replied:
I have already registered on behalf of the organization. Please use the code provided

sandeep-v1404 replied:
Yes 

> I have already registered on behalf of the organization. Please use the code provided

But the code is related to `talawa-admin`

palisadoes commented on .github/workflows/pull-request.yml:
```diff
@@ -70,14 +70,26 @@ jobs:
   Code-Coverage:
     name: Shows code coverage
     runs-on: ubuntu-latest
-    needs: CodeQL
```
This needs to be restored. We need Code-Coverage to always be run after CodeQL

palisadoes commented on .github/workflows/push.yml:
```diff
@@ -26,3 +26,30 @@ jobs:
 
        - name: Perform CodeQL Analysis
          uses: github/codeql-action/analyze@v1
+         
```
We don't need to run the code coverage report on pushes. Please revert all changes to this file.

palisadoes reviewed (changes requested):
@sandeep-v1404 Please make the suggested changes. This PR will really help us to improve our code quality.

palisadoes commented on constants.js:
```diff
@@ -1,4 +1,4 @@
-let URL = 'http://calico.palisadoes.org/talawa/graphql';
+let URL = 'https://talawa-graphql-api.herokuapp.com/graphql';
```
Please revert this. You are right in that it is incorrect, but we should limit the changes to fixing test coverage. Please open another issue to fix this and I will assign it to you.

palisadoes reviewed (changes requested):
We are almost there. Please make the changes and let's get these checks going!

palisadoes commented:
Use this badge:
[![codecov](https://codecov.io/gh/PalisadoesFoundation/talawa-api/branch/master/graph/badge.svg?token=CECBQTAOKM)](https://codecov.io/gh/PalisadoesFoundation/talawa-api)

`[![codecov](https://codecov.io/gh/PalisadoesFoundation/talawa-api/branch/master/graph/badge.svg?token=CECBQTAOKM)](https://codecov.io/gh/PalisadoesFoundation/talawa-api)`

palisadoes commented:
Please submit a test PR to confirm whether this is now working.

palisadoes commented:
Sandeep I created the PR to do the test. The dashboard shows our API code coverage! What I don't understand is that it states that we have 59.02% coverage but the `coverage sunburst` graphic doesn't show this, it shows much less. What is the GitHub action reading incorrectly, or is it? Do you know why?

![image](https://user-images.githubusercontent.com/16875803/148480306-1af69cd8-a2b5-44e4-97e8-ac170c7d835a.png)

Here is the reference link: https://app.codecov.io/gh/PalisadoesFoundation/talawa-api

palisadoes commented:
I also think we the GitHub action should fail if the tests fail. The line in the [pull-request.yml](https://github.com/PalisadoesFoundation/talawa-api/blob/develop/.github/workflows/pull-request.yml) file should not have this line:

`      continue-on-error: true`

Is this the correct command to run for testing?

`      run: npm run test-dev`

sandeep-v1404 commented:
> Sandeep I created the PR to do the test. The dashboard shows our API code coverage! What I don't understand is that it states that we have 59.02% coverage but the `coverage sunburst` graphic doesn't show this, it shows much less. What is the GitHub action reading incorrectly, or is it? Do you know why?
> 
> ![image](https://user-images.githubusercontent.com/16875803/148480306-1af69cd8-a2b5-44e4-97e8-ac170c7d835a.png)
> 
> Here is the reference link: https://app.codecov.io/gh/PalisadoesFoundation/talawa-api

Actually, the URL in constants which we reverted is not working. And previously when I tested with that URL, the test coverage was 67%. Since we changed the URL, the test coverage percentage decreased which lead to action fail.

sandeep-v1404 commented:
> I also think we the GitHub action should fail if the tests fail. The line in the [pull-request.yml](https://github.com/PalisadoesFoundation/talawa-api/blob/develop/.github/workflows/pull-request.yml) file should not have this line:
> 
> ` continue-on-error: true`
> 
> Is this the correct command to run for testing?
> 
> ` run: npm run test-dev`

Sir, there are some tests failing currently, so for testing purpose I kept `continue-on-error: true` as it is. We can remove it later when all the tests updated and run successfully. And that's the right command to run testing sir.

palisadoes commented:
Thanks, but if the tests are failing, then the PR should not not pass. We are aiming for better code quality. 

@sumitra19jha please comment on this. I'm concerned that we are creating a strict testing regime that isn't strict.
